File: 백준/Platinum/11438. LCA 2/LCA 2.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    private static List<List<Integer>> tree;
    private static int[] depth;
    private static int[][] parent;
    private static boolean[] visited;
    private static int maxLevel;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());

        tree = new ArrayList<>();
        for (int i = 0; i <= n; i++) {
            tree.add(new ArrayList<>());
        }

        for (int i = 0; i < n - 1; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            tree.get(a).add(b);
            tree.get(b).add(a);
        }

        maxLevel = 0;
        int tmp = 1;
        while (tmp <= n) {
            tmp <<= 1;
            maxLevel++;
        }

        depth = new int[n + 1];
        parent = new int[maxLevel + 1][n + 1];
        visited = new boolean[n + 1];

        bfs(1);

        for (int k = 1; k <= maxLevel; k++) {
            for (int node = 1; node <= n; node++) {
                parent[k][node] = parent[k - 1][parent[k - 1][node]];
            }
        }

        StringBuilder sb = new StringBuilder();
        int m = Integer.parseInt(br.readLine());
        for (int i = 0; i < m; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int lca = findLCA(a, b);
            sb.append(lca).append('\n');
        }

        System.out.print(sb);
    }

    private static void bfs(int root) {
        Queue<Integer> queue = new LinkedList<>();
        queue.offer(root);
        visited[root] = true;
        depth[root] = 0;
        parent[0][root] = 0;

        while (!queue.isEmpty()) {
            int cur = queue.poll();

            for (int next : tree.get(cur)) {
                if (!visited[next]) {
                    visited[next] = true;
                    depth[next] = depth[cur] + 1;
                    parent[0][next] = cur;
                    queue.offer(next);
                }
            }
        }
    }

    private static int findLCA(int a, int b) {
        if (depth[a] > depth[b]) {
            int tmp = a;
            a = b;
            b = tmp;
        }

        int diff = depth[b] - depth[a];
        for (int k = 0; k <= maxLevel; k++) {
            if ((diff & (1 << k)) != 0) {
                b = parent[k][b];
            }
        }

        if (a == b) return a;

        for (int k = maxLevel; k >= 0; k--) {
            if (parent[k][a] != parent[k][b]) {
                a = parent[k][a];
                b = parent[k][b];
            }
        }

        return parent[0][a];
    }
}

File: 백준/Platinum/11438. LCA 2/LCA 2.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.LinkedList
import java.util.Queue

private lateinit var tree: MutableList<MutableList<Int>>
private lateinit var depth: IntArray
private lateinit var parent: Array<IntArray>
private lateinit var visited: BooleanArray
private var maxLevel = 0

fun main() {
    val br = BufferedReader(InputStreamReader(System.`in`))
    val n = br.readLine().toInt()

    tree = MutableList(n + 1) { mutableListOf() }

    repeat(n - 1) {
        val (a, b) = br.readLine().split(" ").map { it.toInt() }
        tree[a].add(b)
        tree[b].add(a)
    }

    maxLevel = 0
    var tmp = 1
    while (tmp <= n) {
        tmp = tmp shl 1
        maxLevel++
    }

    depth = IntArray(n + 1)
    parent = Array(maxLevel + 1) { IntArray(n + 1) }
    visited = BooleanArray(n + 1)

    bfs(1)

    for (k in 1..maxLevel) {
        for (node in 1..n) {
            parent[k][node] = parent[k - 1][parent[k - 1][node]]
        }
    }

    val sb = StringBuilder()
    val m = br.readLine().toInt()
    repeat(m) {
        val (a, b) = br.readLine().split(" ").map { it.toInt() }
        val lca = findLCA(a, b)
        sb.append(lca).append('\n')
    }

    print(sb)
}

private fun bfs(root: Int) {
    val queue: Queue<Int> = LinkedList()
    queue.offer(root)
    visited[root] = true
    depth[root] = 0
    parent[0][root] = 0

    while (queue.isNotEmpty()) {
        val cur = queue.poll()

        for (next in tree[cur]) {
            if (!visited[next]) {
                visited[next] = true
                depth[next] = depth[cur] + 1
                parent[0][next] = cur
                queue.offer(next)
            }
        }
    }
}

private fun findLCA(a: Int, b: Int): Int {
    var nodeA = a
    var nodeB = b

    if (depth[nodeA] > depth[nodeB]) {
        nodeA = nodeB.also { nodeB = nodeA }
    }

    val diff = depth[nodeB] - depth[nodeA]
    for (k in 0..maxLevel) {
        if ((diff and (1 shl k)) != 0) {
            nodeB = parent[k][nodeB]
        }
    }

    if (nodeA == nodeB) return nodeA

    for (k in maxLevel downTo 0) {
        if (parent[k][nodeA] != parent[k][nodeB]) {
            nodeA = parent[k][nodeA]
            nodeB = parent[k][nodeB]
        }
    }

    return parent[0][nodeA]
}
